packages/core/src/components/ResponseErrorPanel/ResponseErrorPanel.tsx

```diff
@@ -39,87 +39,119 @@ const useStyles = makeStyles(theme => ({
-  const requestString = request && `${request.method} ${request.url}`;
-  const messageString = cause.message.replace(/\\n/g, '\n');
-  const stackString = cause.stack?.replace(/\\n/g, '\n');
-  const jsonString = JSON.stringify(data, undefined, 2);
-
   return (
     <List dense>
       <ListItem alignItems="flex-start">
         <ListItemText
           classes={{ secondary: classes.text }}
           primary="Error"
-          secondary={errorString}
+          secondary={error}
         />
-        <CopyTextButton text={errorString} />
+        <CopyTextButton text={error} />
       </ListItem>
       <ListItem alignItems="flex-start">
         <ListItemText
           classes={{ secondary: classes.text }}
           primary="Message"
-          secondary={messageString}
+          secondary={message}
         />
-        <CopyTextButton text={messageString} />
+        <CopyTextButton text={message} />
       </ListItem>
-      {requestString && (
+      {request && (
         <ListItem alignItems="flex-start">
           <ListItemText
             classes={{ secondary: classes.text }}
             primary="Request"
-            secondary={requestString}
+            secondary={request}
           />
-          <CopyTextButton text={requestString} />
+          <CopyTextButton text={request} />
         </ListItem>
       )}
-      {stackString && (
+      {stack && (
         <ListItem alignItems="flex-start">
           <ListItemText
             classes={{ secondary: classes.text }}
             primary="Stack Trace"
-            secondary={stackString}
+            secondary={stack}
           />
-          <CopyTextButton text={stackString} />
+          <CopyTextButton text={stack} />
         </ListItem>
       )}
-      <Divider component="li" className={classes.divider} />
-      <ListItem alignItems="flex-start">
-        <ListItemText
-          classes={{ secondary: classes.text }}
-          primary="Full Error as JSON"
-          secondary={<CodeSnippet language="json" text={jsonString} />}
-        />
-        <CopyTextButton text={jsonString} />
-      </ListItem>
+      {json && (
+        <>
+          <Divider component="li" className={classes.divider} />
+          <ListItem alignItems="flex-start">
+            <ListItemText
+              classes={{ secondary: classes.text }}
+              primary="Full Error as JSON"
+              secondary={<CodeSnippet language="json" text={json} />}
+            />
+            <CopyTextButton text={json} />
+          </ListItem>
+        </>
+      )}
     </List>
   );
 };
 
+type Props = {
+  error: Error;
+};
+
+/**
+ * Renders details about a failed server request.
+ *
+ * Has special treatment for ResponseError errors, to display rich
+ * server-provided information about what happened.
+ */
+export const ResponseErrorDetails = ({ error }: Props) => {
+  if (error.name !== 'ResponseError') {
+    return (
+      <ResponseErrorList
```

Comment:
This is really the bulk of the change, the rest of it just extracted it up to a toplevel component.